src/sqlfluff/core/parser/segments/base.py

```diff
@@ -782,9 +782,18 @@ def match(
                 )
             # Once unified we can deal with it just as a MatchResult
             if m.has_match():
-                return MatchResult(
-                    (cls(segments=m.matched_segments),), m.unmatched_segments
-                )
+                try:
+                    return MatchResult(
+                        (cls(segments=m.matched_segments),), m.unmatched_segments
+                    )
+                except TypeError as err:  # pragma: no cover
+                    # This is an error to assist with debugging dialect design.
+                    # It's most likely that the match_grammar has been seg on
+                    # a raw segment which shouldn't happen.
+                    raise TypeError(
+                        f"Error in instantiating {cls.__module__}.{cls.__name__}. Have "
+                        f"you defined a match_grammar on a RawSegment? : {str(err)}"
+                    )
```

Comment:
```suggestion
                    ) from err
```
This will preserve the original exception (which I think will appear in the stack backtrace).
https://stackoverflow.com/questions/24752395/python-raise-from-usage

I haven't tested this, so probably worth a quick manual test if you accept this suggestion.

Reply:
I think I'm going to leave this out. In some ways the stack trace is actively unhelpful for this error - it will just have a big old trace through a load of parsing code (in a very nested way). I kind of wanted to actually _hide_ the stack trace but that seemed too hard.